baby-liminal-extension/build.rs

```diff
@@ -0,0 +1,247 @@
+        let pk = keygen_pk(params, vk.clone(), &circuit).expect("pk should not fail");
+
+        let mut transcript = Blake2bWrite::<_, _, Challenge255<_>>::init(vec![]);
+        create_proof::<_, ProverGWC<'_, Bn256>, _, _, _, _>(
+            params,
+            &pk,
+            &[circuit],
+            &[&[&instances]],
+            ParamsKZG::<Bn256>::mock_rng(),
+            &mut transcript,
+        )
+        .expect("prover should not fail");
+
+        Artifacts {
+            verification_key: serialize_vk(vk, params.k()),
+            proof: transcript.finalize(),
+            public_input: instances.iter().flat_map(|i| i.to_bytes()).collect(),
+        }
+    }
+
+    /// Serializes the verification key to the raw bytes together with the upperbound on the number of gates (required
+    /// by the on-chain verifier).
+    fn serialize_vk(vk: VerifyingKey<G1Affine>, k: u32) -> Vec<u8> {
+        let mut buffer = Vec::new();
+        buffer.extend(k.to_le_bytes());
+        buffer.extend(vk.to_bytes(halo2_proofs::SerdeFormat::RawBytesUnchecked));
+        buffer
+    }
+}
+
+/// This module defines the circuit for the benchmarking purposes.
+#[cfg(feature = "runtime-benchmarks")]
+mod circuit {
+    use halo2_proofs::{
+        circuit::{Layouter, Region, Value},
+        halo2curves::bn256::Fr,
+        plonk::{Circuit, ConstraintSystem, Error},
+        standard_plonk::{StandardPlonk, StandardPlonkConfig},
+    };
+
+    pub struct BenchCircuit<const INSTANCES: usize, const ROW_BLOWUP: usize> {
+        /// The roots of the instances (`i`th root is a square root of `i`th instance).
+        roots: [Fr; INSTANCES],
+    }
+
+    impl<const INSTANCES: usize, const ROW_BLOWUP: usize> Default
+        for BenchCircuit<INSTANCES, ROW_BLOWUP>
+    {
+        fn default() -> Self {
+            BenchCircuit {
+                roots: [Fr::zero(); INSTANCES],
+            }
+        }
+    }
+
+    impl<const INSTANCES: usize, const ROW_BLOWUP: usize> BenchCircuit<INSTANCES, ROW_BLOWUP> {
+        /// Create a circuit with the consecutive natural numbers as advices.
+        pub fn natural_numbers() -> Self {
+            let roots: Vec<_> = (0..INSTANCES).map(|i| Fr::from(i as u64)).collect();
+            Self {
+                roots: roots.try_into().unwrap(),
+            }
+        }
+
+        /// Assign the `idx`th root to the `a` and `b` advices and `-1` to the `q_ab` fixed value.
+        ///
+        /// We assume here that this is the first row in the region (offset 0).
+        fn neg_root_square(
+            &self,
+            idx: usize,
+            region: &mut Region<Fr>,
+            config: &StandardPlonkConfig<Fr>,
+        ) -> Result<(), Error> {
+            region.assign_advice(|| "", config.a, 0, || Value::known(self.roots[idx]))?;
+            region.assign_advice(|| "", config.b, 0, || Value::known(self.roots[idx]))?;
+            region.assign_fixed(|| "", config.q_ab, 0, || Value::known(-Fr::one()))?;
```

Comment:
these names seem to be usefull when debugging - call it e.g. -1

Reply:
already covered